### PR TITLE
Add framework Engine API client

### PR DESCRIPTION
**Pull request by GGP1**
|Related issue|
|---|
| Closes https://github.com/wazuh/wazuh/issues/24646 |

## Description

Adds the `Engine` class to communicate to the Engine API via HTTP over Unix sockets.

## Tests

<details><summary>Unit tests</summary>

```console
(venv) gasti@gasti:~/work/wazuh$ pytest framework/wazuh/core/engine --disable-warnings
============================================================== test session starts ===============================================================
platform linux -- Python 3.10.14, pytest-7.3.1, pluggy-1.5.0
rootdir: /home/gasti/work/wazuh/framework
configfile: pytest.ini
plugins: asyncio-0.18.1, tavern-1.23.5, html-2.1.1, metadata-3.1.1, cov-4.1.0, anyio-4.1.0, trio-0.8.0, aiohttp-1.0.4
asyncio: mode=auto
collected 10 items                                                                                                                               

framework/wazuh/core/engine/tests/test_base.py .                                                                                           [ 10%]
framework/wazuh/core/engine/tests/test_init.py .........                                                                                   [100%]

=============================================================== 10 passed in 0.28s ===============================================================
```

</details>

<details><summary>Test using the Docker API as the server</summary>

> The Engine API is still unavailable, so I tested the class against the Docker API which uses the same protocols.

```console
(venv) gasti@gasti:~/work/wazuh$ /home/gasti/work/wazuh/venv/bin/python3 -m asyncio
asyncio REPL 3.10.14 (main, May  3 2024, 08:57:23) [GCC 12.2.0] on linux
Use "await" directly instead of "asyncio.run()".
Type "help", "copyright", "credits" or "license" for more information.
>>> import asyncio
>>> from wazuh.core.engine import Engine
>>> engine = Engine(socket_path='/var/run/docker.sock')
>>> response = await engine._client.get("http://docker/info")
>>> print(response)
<Response [200 OK]>
>>> print(response.json())
{'ID': '***', 'Containers': 0, 'ContainersRunning': 0, 'ContainersPaused': 0, 'ContainersStopped': 0, ...}
```

</details>

### CODE REVIEW

File: framework/wazuh/core/engine/__init__.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+from contextlib import asynccontextmanager
+from logging import getLogger
+from typing import AsyncIterator
+
+from httpx import AsyncClient, AsyncHTTPTransport, ConnectError, Timeout, TimeoutException, UnsupportedProtocol
+from wazuh.core.exception import WazuhEngineError
+
+logger = getLogger('wazuh')
+
+# TODO: use actual default path
```
Comment (GGP1) [inline framework/wazuh/core/engine/__init__.py:10]: I'm making a list with these TODOs so we can create a technical debt issue to solve them later

**Review by RamosFe — APPROVED** [review]
LGTM! 🚀

File: framework/wazuh/core/engine/tests/test_init.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    'params',
+    [
+        {'retries': 3, 'timeout': 10},
+        {'retries': None, 'timeout': 15},
+        {'timeout': 0},
+        {},
+    ],
+)
+def test_engine_init(params: dict):
+    """Check the correct initalization of the `Engine` class."""
```
Comment (fdalmaup) [inline framework/wazuh/core/engine/tests/test_init.py:19]:
```suggestion
    """Check the correct initialization of the `Engine` class."""
```

File: framework/wazuh/core/engine/tests/test_base.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+from unittest import mock
+
+from wazuh.core.engine.base import BaseModule
+
+
+def test_base_module_init():
+    """Check the correct initalization of the `BaseModule` class."""
```
Comment (fdalmaup) [inline framework/wazuh/core/engine/tests/test_base.py:7]:
```suggestion
    """Check the correct initialization of the `BaseModule` class."""
```

File: framework/wazuh/core/engine/tests/test_init.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        assert engine._client._transport._pool._retries == DEFAULT_RETRIES
+
+    if 'timeout' in params:
+        assert engine._client.timeout == Timeout(params['timeout'])
+    else:
+        assert engine._client.timeout == Timeout(DEFAULT_TIMEOUT)
+
+
+@pytest.mark.asyncio
+async def test_close():
```
Comment (fdalmaup) [inline framework/wazuh/core/engine/tests/test_init.py:37]:
```suggestion
async def test_engine_close():
```

File: framework/wazuh/core/engine/tests/test_init.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        {'timeout': 0},
+        {},
+    ],
+)
+def test_engine_init(params: dict):
+    """Check the correct initalization of the `Engine` class."""
+    engine = Engine(socket_path='/test.sock', **params)
+
+    assert isinstance(engine._client, AsyncClient)
+    assert engine._client.is_closed == False
```
Comment (fdalmaup) [inline framework/wazuh/core/engine/tests/test_init.py:23]:
```suggestion
    assert not engine._client.is_closed
```

File: framework/wazuh/core/engine/tests/test_init.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    engine._client.aclose.assert_called_once()
+
+
+@pytest.mark.asyncio
+async def test_get_engine_client():
+    """Check the correct behavior of the `get_engine_client` function."""
+    async with get_engine_client() as engine:
+        assert engine._client.is_closed == False
+
+    assert engine._client.is_closed == True
```
Comment (fdalmaup) [inline framework/wazuh/core/engine/tests/test_init.py:52]:
```suggestion
        assert not engine._client.is_closed

    assert engine._client.is_closed
```

**Review by fdalmaup — CHANGES_REQUESTED** [review]

**Review by fdalmaup — APPROVED** [review]
LGTM

**Pull request merged**